src/psyclone/dynamo0p3.py

```diff
@@ -7909,16 +7910,23 @@ def colourmap(self):
         if not self.is_coloured():
             raise InternalError(f"Kernel '{self.name}' is not inside a "
                                 f"coloured loop.")
+        sched = self.ancestor(InvokeSchedule)
         if self._is_intergrid:
-            invoke = self.ancestor(InvokeSchedule).invoke
+            invoke = sched.invoke
             if id(self) not in invoke.meshes.intergrid_kernels:
                 raise InternalError(
                     f"Colourmap information for kernel '{self.name}' has "
                     f"not yet been initialised")
             cmap = invoke.meshes.intergrid_kernels[id(self)].\
                 colourmap_symbol.name
         else:
-            cmap = self.scope.symbol_table.lookup_with_tag("cmap").name
+            try:
+                cmap = sched.symbol_table.lookup_with_tag("cmap").name
+            except KeyError:
+                # We have to do this here as _init_colourmap is only called
+                # at code-generation time.
+                cmap = sched.symbol_table.find_or_create_array(
```

Comment:
I'm a bit confused here. I thought colourmap information was added at gen() time on purpose so that all transformations had completed. If so, why do we need the information before then (which I guess is what is happening in the except).. Am I right in thinking that the exception on line 7917 is the equivalent of this KeyError?

Reply:
I agree it is confusing. However, this method *is* called at code-generation time. I've updated the comment. Yes, 7917 is the equivalent of this situation for intergrid kernels.

Reply:
Thanks for clarifying